api_callable.go

```diff
@@ -0,0 +1,81 @@
+package gax
+
+import (
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	retryableGRPCCodes = map[codes.Code]bool{
+		codes.DeadlineExceeded: true,
+		codes.Aborted:          true,
+	}
+)
+
+// Represents a GRPC call stub.
+type APICall func(context.Context, interface{}) (interface{}, error)
+
+// scaleDuration returns the product of a and mult.
+func scaleDuration(a time.Duration, mult float64) time.Duration {
+	ns := float64(a) * mult
+	return time.Duration(int64(ns))
+}
+
+// stubWithRetry returns a wrapper for stub with an exponential backoff retry
+// mechanism based on the values provided in backoffSettings.
+func stubWithRetry(ctx context.Context, stub APICall, backoffSettings backoffSettings) APICall {
+	return func(ctx context.Context, req interface{}) (resp interface{}, err error) {
+		// Forces ctx to expire after a deadline.
+		ctx, _ = context.WithTimeout(ctx, backoffSettings.totalTimeout)
+
+		delay := backoffSettings.delayTimeoutSettings.initial
+		timeout := backoffSettings.rpcTimeoutSettings.initial
+
+		for {
+			// If the deadline is exceeded...
+			if ctx.Err() != nil {
+				return nil, ctx.Err()
+			}
+			timeoutCtx, _ := context.WithTimeout(ctx, backoffSettings.rpcTimeoutSettings.max)
+			timeoutCtx, _ = context.WithTimeout(timeoutCtx, timeout)
+			resp, err = stub(timeoutCtx, req)
+			code := grpc.Code(err)
+			if code == codes.OK {
+				break
+			}
+			if !retryableGRPCCodes[code] {
```

Comment:
not using retry_codes in the config -- are you going to do it in a different patch?